File: tests/cpp/test_Cell.cpp
#include <sdot/Cell.h>
#include "catch_main.h"

TEST_CASE( "Cell", "" ) {
    Cell<Config> cell;
    cell.init_geometry_to_encompass( { FromItemValue(), -2 }, { FromItemValue(), 2 } );
    cell.p0 = { 0, 0 };
    cell.w0 = 0;
    cell.i0 = 0;

    // for( PI d = 0; d < nb_dims; ++d ) {
    //     Point p( FromItemValue(), 0 ); p[ d ] = +1;
    //     Point q( FromItemValue(), 0 ); q[ d ] = -1;
    //     cell.cut_boundary( p, 1, 2 * d + 0 );
    //     cell.cut_boundary( q, 0, 2 * d + 1 );
    // }

    cell.cut_boundary( { -1, 0 }, 0, 1 );
    cell.cut_boundary( { 0, -1 }, 0, 1 );
    cell.cut_boundary( { 1, 1 }, 1, 1 );
    cell.memory_compaction();
    P( cell );

    // if ( compaction )
    //     cell.memory_compaction();
    P( cell.for_each_cut_with_measure( [&]( const CellCut<Config> &cut, double measure ) {
        P( cut, measure );
    } ) );

    P( cell.measure() );

    // VtkOutput vo;
    // cell.display_vtk( vo );
    // vo.save( "out.vtk" );
}

File: tests/python/test_Cell.py
from sdot import Cell
import numpy

def test_Cell_emptyness_bounded():
    c = Cell( ndim = 2 )
    assert c.nb_vertices == 0
    assert c.bounded == False
    assert c.empty == False

    c.cut_boundary( [ +1, 0 ], 1 )
    assert c.nb_vertices == 0
    assert c.bounded == False
    assert c.empty == False

    c.cut_boundary( [ 0, +1 ], 1 )
    assert c.nb_vertices == 1
    assert c.bounded == False
    assert c.empty == False

    c.cut_boundary( [ -1, 0 ], 1 )
    assert c.nb_vertices == 2
    assert c.bounded == False
    assert c.empty == False

    c.cut_boundary( [ 0, -1 ], 1 )
    assert c.nb_vertices == 4
    assert c.bounded == True
    assert c.empty == False

    c.cut_boundary( [ 1, 0 ], 0.5 )
    assert c.nb_vertices == 4
    assert c.bounded == True
    assert c.empty == False

    c.cut_boundary( [ 1, 0 ], -2 )
    assert c.nb_vertices == 0
    assert c.bounded == True
    assert c.empty == True

    c.cut_boundary( [ 1, 0 ], -3 )
    assert c.nb_vertices == 0
    assert c.bounded == True
    assert c.empty == True

def test_Cell_emptyness_unbounded():
    c = Cell( ndim = 3 )
    assert c.nb_vertices == 0
    assert c.bounded == False
    assert c.empty == False

    c.cut_boundary( [ 1, 1, 0 ], 1 )
    print( "nb_vertices", c.nb_vertices )
    print( "bounded", c.bounded )
    print( "empty", c.empty )

    c.cut_boundary( [ -1, 0, 0 ], 0 )
    print( "nb_vertices", c.nb_vertices )
    print( "bounded", c.bounded )
    print( "empty", c.empty )

    c.cut_boundary( [ 0, -1, 0 ], 0 )
    print( "nb_vertices", c.nb_vertices )
    print( "bounded", c.bounded )
    print( "empty", c.empty )

    # assert c.nb_vertices == 0
    # assert c.bounded == False
    # assert c.empty == False

    # c.cut_boundary( [ 0, -1 ], 0 )
    # assert c.nb_vertices == 1
    # assert c.bounded == False
    # assert c.empty == True

# test_Cell_emptyness_bounded()
test_Cell_emptyness_unbounded()

# cut( [ +1, 0, 0 ] )
# cut( [ 0, +1, 0 ] )
# cut( [ 0, 0, +1 ] )
# cut( [ -1, 0, 0 ] )
# cut( [ 0, -1, 0 ] )
# cut( [ 0, 0, -1 ] )

# print( cell )

# vo = module.VtkOutput()
# cell.display_vtk( vo )
# vo.save( "pd.vtk" )
